File: backend/app/services/interest.py
"""Tiered interest calculation service (charter §4.5, §5).

B account: 3-tier monthly interest
C account: 5% annual, monthly dividend to A
"""

from decimal import ROUND_DOWN, Decimal


def calculate_c_dividend(c_balance_cents: int, annual_rate_bps: int) -> int:
    """Calculate C account monthly dividend.

    C_balance × annual_rate / 10000 / 12, floored to cents.

    Args:
        c_balance_cents: C balance in cents
        annual_rate_bps: Annual rate in basis points (万分比), e.g. 500 = 5.0%

    Returns:
        Dividend amount in cents (floored)
    """
    if c_balance_cents <= 0:
        return 0

    balance = Decimal(c_balance_cents)
    rate = Decimal(annual_rate_bps) / Decimal(10000)
    monthly = balance * rate / Decimal(12)
    return int(monthly.to_integral_value(rounding=ROUND_DOWN))


def calculate_b_interest(
    b_principal: int,
    p_active: int,
    tier1_rate: int,
    tier1_limit: int,
    tier2_rate: int,
    tier3_rate: int,
    is_suspended: bool = False,
) -> dict[str, int]:
    """Calculate B account tiered monthly interest.

    Tier1: 0 ~ tier1_limit (default 1000元) at tier1_rate (default 2.0%)
    Tier2: tier1_limit ~ p_active at tier2_rate (default 1.2%)
    Tier3: p_active ~ 1.2×p_active at tier3_rate (default 0.3%)

    All rates are in basis points (万分比): 200 = 2.0%

    Args:
        b_principal: B principal balance in cents
        p_active: Current target price in cents
        tier1_rate: Tier 1 monthly rate in bps
        tier1_limit: Tier 1 upper limit in cents
        tier2_rate: Tier 2 monthly rate in bps
        tier3_rate: Tier 3 monthly rate in bps
        is_suspended: If True, return all zeros

    Returns:
        Dict with tier1, tier2, tier3, total (all in cents)
    """
    if is_suspended or b_principal <= 0:
        return {"tier1": 0, "tier2": 0, "tier3": 0, "total": 0}

    # When no wish list (p_active=0), only Tier1 applies (no Tier2/3 range)
    if p_active <= 0:
        tier1_base = min(b_principal, tier1_limit)
        tier1 = _calc_tier(tier1_base, tier1_rate)
        return {"tier1": tier1, "tier2": 0, "tier3": 0, "total": tier1}

    cap_overflow = int(Decimal(p_active) * Decimal("1.2"))

    # Tier 1: 0 ~ tier1_limit
    tier1_base = min(b_principal, tier1_limit)
    tier1 = _calc_tier(tier1_base, tier1_rate)

    # Tier 2: tier1_limit ~ p_active
    tier2_base = max(0, min(b_principal, p_active) - tier1_limit)
    tier2 = _calc_tier(tier2_base, tier2_rate)

    # Tier 3: p_active ~ cap_overflow
    tier3_base = max(0, min(b_principal, cap_overflow) - p_active)
    tier3 = _calc_tier(tier3_base, tier3_rate)

    total = tier1 + tier2 + tier3

    return {"tier1": tier1, "tier2": tier2, "tier3": tier3, "total": total}


def _calc_tier(base_cents: int, rate_bps: int) -> int:
    """Calculate interest for a single tier: base × rate / 10000, floored."""
    if base_cents <= 0 or rate_bps <= 0:
        return 0
    result = Decimal(base_cents) * Decimal(rate_bps) / Decimal(10000)
    return int(result.to_integral_value(rounding=ROUND_DOWN))

File: cloudfunctions/_shared/interest.js
'use strict';

/**
 * Calculate C account monthly dividend (C -> A)
 * Formula: floor(C_balance * annual_rate / 10000 / 12)
 * 
 * @param {BigInt} cBalance - C account balance in cents
 * @param {number} annualRateBps - Annual rate in basis points (500 = 5.0%)
 * @returns {BigInt} dividend amount in cents
 */
function calculateCDividend(cBalance, annualRateBps) {
  const balance = BigInt(cBalance);
  const rate = BigInt(annualRateBps);
  if (balance <= 0n) return 0n;
  return balance * rate / 10000n / 12n;
}

/**
 * Calculate B account tiered interest
 * Tier1: min(B, tier1_limit) * tier1_rate / 10000
 * Tier2: max(0, min(B, P_active) - tier1_limit) * tier2_rate / 10000
 * Tier3: max(0, min(B, 1.2*P_active) - P_active) * tier3_rate / 10000
 *
 * @param {BigInt} bPrincipal - B account balance (principal only) in cents
 * @param {BigInt} pActive - P_active value in cents (target price or max_price)
 * @param {Object} rates - { tier1Rate, tier1Limit, tier2Rate, tier3Rate } all in bps/cents
 * @param {boolean} isSuspended - Whether interest is suspended
 * @returns {{ tier1: BigInt, tier2: BigInt, tier3: BigInt, total: BigInt }}
 */
function calculateBInterest(bPrincipal, pActive, rates, isSuspended) {
  const b = BigInt(bPrincipal);
  const p = BigInt(pActive);
  const tier1Rate = BigInt(rates.tier1Rate);
  const tier1Limit = BigInt(rates.tier1Limit);
  const tier2Rate = BigInt(rates.tier2Rate);
  const tier3Rate = BigInt(rates.tier3Rate);

  if (isSuspended || b <= 0n) {
    return { tier1: 0n, tier2: 0n, tier3: 0n, total: 0n };
  }

  // Cap overflow = 1.2 * P_active
  const capOverflow = p * 12n / 10n;

  // Tier 1: 0 ~ tier1_limit
  const tier1Base = b < tier1Limit ? b : tier1Limit;
  const tier1 = tier1Base * tier1Rate / 10000n;

  let tier2 = 0n;
  let tier3 = 0n;

  if (p > 0n) {
    // Tier 2: tier1_limit ~ P_active
    const tier2Upper = b < p ? b : p;
    const tier2Base = tier2Upper > tier1Limit ? tier2Upper - tier1Limit : 0n;
    tier2 = tier2Base * tier2Rate / 10000n;

    // Tier 3: P_active ~ 1.2*P_active
    const tier3Upper = b < capOverflow ? b : capOverflow;
    const tier3Base = tier3Upper > p ? tier3Upper - p : 0n;
    tier3 = tier3Base * tier3Rate / 10000n;
  }

  return { tier1, tier2, tier3, total: tier1 + tier2 + tier3 };
}

module.exports = { calculateCDividend, calculateBInterest };
